### PR TITLE
DM-47947: add functions for processing concatenations of encoded regions

### PR DESCRIPTION
**Pull request by TallJimbo**
## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: include/lsst/sphgeom/CompoundRegion.h [hunk 1]
```diff
@@ -95,11 +97,21 @@ class CompoundRegion : public Region {
...
     // Implementation helper for decode().
-    static std::array<std::unique_ptr<Region>, 2> _decode(
+    static std::vector<std::unique_ptr<Region>> _decode(
         std::uint8_t tc, std::uint8_t const *buffer, std::size_t nBytes);
 
+    // Provide read access to all operands for quick iteration.
+    std::vector<std::unique_ptr<Region>> const& operands() const { return _operands; }
+
+    // Check if we have any operands.
+    bool empty() const { return _operands.empty(); }
```
Comment (TallJimbo) [inline include/lsst/sphgeom/CompoundRegion.h:107]: Given that `Region` also has an `isEmpty` method that's about whether there are any points in the region, I think this might be more confusing than helpful.

File: src/CompoundRegion.cc [hunk 2]
```diff
@@ -61,28 +62,53 @@ std::uint64_t consumeDecodeU64(std::uint8_t const *&buffer, std::uint8_t const *
 template <typename F>
 auto getUnionBounds(UnionRegion const &compound, F func) {
     auto bounds = func(compound.getOperand(0));
-    bounds.expandTo(func(compound.getOperand(1)));
+    for (unsigned i = 1; i < compound.nOperands(); ++ i) {
```
Comment (TallJimbo) [inline src/CompoundRegion.cc:65]:
```suggestion
    for (std::size_t i = 1; i < compound.nOperands(); ++ i) {
```

File: src/CompoundRegion.cc [hunk 3]
```diff
@@ -61,28 +62,53 @@ std::uint64_t consumeDecodeU64(std::uint8_t const *&buffer, std::uint8_t const *
...
+        bounds.expandTo(func(compound.getOperand(i)));
+    }
     return bounds;
 }
 
 template <typename F>
 auto getIntersectionBounds(IntersectionRegion const &compound, F func) {
     auto bounds = func(compound.getOperand(0));
-    bounds.clipTo(func(compound.getOperand(1)));
+    for (unsigned i = 1; i < compound.nOperands(); ++ i) {
```
Comment (TallJimbo) [inline src/CompoundRegion.cc:74]:
```suggestion
    for (std::size_t i = 1; i < compound.nOperands(); ++ i) {
```

File: include/lsst/sphgeom/CompoundRegion.h (hunk 1 above)
Comment (andy-slac) [reply include/lsst/sphgeom/CompoundRegion.h:107]: OK, I'll drop it, I can use `nOperands()` which is public and will likely produce the same code.

File: src/Region.cc [hunk 4]
```diff
@@ -73,15 +90,63 @@ std::unique_ptr<Region> Region::decode(std::uint8_t const * buffer, size_t n) {
...
+            region_begin = region_end;
+            ++region_begin;
+            region_end = std::find(region_begin, s.end(), ':');
+        }
+        auto bytes = base64::decode_into<std::vector<std::uint8_t>>(region_begin, region_end);
+        union_args.push_back(decode(bytes));
+        return std::unique_ptr<UnionRegion>(new UnionRegion(std::move(union_args)));
+    } else {
+        auto bytes = base64::decode_into<std::vector<std::uint8_t>>(region_begin, region_end);
+        return decode(bytes);
```
Comment (andy-slac) [inline src/Region.cc:110]: My guess is that in this scheme an empty UnionRegion is represented by an empty string? I'm not sure that it could happen in our daf_butler queries, but it may be worth to support it in general, and this code will probably throw an exception for an empty string?

File: tests/test_CompoundRegion.py [hunk 5]
```diff
@@ -228,6 +229,18 @@ def testBounding(self):
...
+        union regions.
+        """
+        # Test with the full UnionRegion encoded, then base64-encoded.
+        s1 = b64encode(self.instance.encode()).decode("ascii")
+        self.assertCompoundRegionsEqual(Region.decodeBase64(s1), self.instance)
+        # Test alternate form with union members concatenated with ':' after
+        # base64-encoding.
+        s2 = ":".join(b64encode(region.encode()).decode("ascii") for region in self.instance)
+        self.assertCompoundRegionsEqual(Region.decodeBase64(s2), self.instance)
+
```
Comment (andy-slac) [inline tests/test_CompoundRegion.py:243]: Maybe add a test with an empty UnionRegion?

File: include/lsst/sphgeom/Region.h [hunk 6]
```diff
@@ -149,9 +166,55 @@ class Region {
     static std::unique_ptr<Region> decode(std::uint8_t const * buffer, size_t n);
     ///@}
 
+    ///@{
+    /// `decodeBase64` deserializes a Region from an ASCII string produced by
+    /// encode and then base64-encoding that result.
+    ///
+    /// This method also interprets ':' as a delimiter for the elements of a
+    /// UnionRegion, to support cases where a union of region is constructed
+    /// server-side in a database as an concatenation with that delimiter.
```
Comment (andy-slac) [inline include/lsst/sphgeom/Region.h:175]: Would it be useful to add support for IntersectionRegion as well, via a different delimiter?

File: tests/test_CompoundRegion.py [hunk 7]
```diff
@@ -321,6 +321,33 @@ def testBounding(self):
...
+        self.assertEqual(run_overlaps([(self.box, self.circle)]), True)
+        self.assertEqual(run_overlaps([(self.box, self.faraway)]), False)
+        self.assertEqual(run_overlaps([(self.circle, self.faraway)]), False)
+        self.assertEqual(run_overlaps([(self.instance, self.box)]), None)
+        self.assertEqual(run_overlaps([(self.box, self.circle), (self.box, self.faraway)]), True)
+        self.assertEqual(run_overlaps([(self.faraway, self.circle), (self.box, self.faraway)]), False)
+        self.assertEqual(run_overlaps([(self.instance, self.box), (self.circle, self.faraway)]), None)
+        self.assertEqual(run_overlaps([(self.instance, self.box), (self.circle, self.box)]), True)
+        self.assertEqual(run_overlaps([(self.circle, self.box), (self.instance, self.box)]), True)
+
```
Comment (andy-slac) [inline tests/test_CompoundRegion.py:350]: Maybe add an exceptional test without `&`?

**Review by andy-slac — APPROVED** [review]
Looks great, few minor comments.

File: include/lsst/sphgeom/Region.h (hunk 6 above)
Comment (TallJimbo) [reply include/lsst/sphgeom/Region.h]: I thought about that too, but I think it'd end up dead code.  And if I'm wrong, it might need to be more complex than allowing just one delimiter or the other (i.e. we'd need to work out associativity).
